packages/Search/src/details/DTK_DetailsTreeConstruction_def.hpp

```diff
@@ -254,32 +255,68 @@ void TreeConstruction<DeviceType>::sortObjects(
     bin_sort.create_permute_vector();
     bin_sort.sort( morton_codes );
-    // TODO: We might be able to just use `bin_sort.get_permute_vector()`
-    // instead of initializing the indices with iota() and sorting the vector
-    bin_sort.sort( object_ids );
     Kokkos::fence();
+
+    // Kokkos::BinSort::get_permute_vector() returns a view with a
+    // value type set to the size type of the memory space.  Cuda memory space
+    // size type is `unsigned int` while it is `size_t` on the host...
+    auto permute = bin_sort.get_permute_vector();
+
+    // The static assertion below is meant to check whether changes occured in
+    // Kokkos_Sort.hpp or in the definition of the size types in which case we
+    // might want to reconsider what to do with the permutation indices.
+    static_assert(
+#ifdef KOKKOS_ENABLE_CUDA
+        std::is_same<typename decltype( permute )::traits::execution_space,
+                     Kokkos::Cuda>::value &&
+                std::is_same<typename decltype( permute )::value_type,
+                             unsigned int>::value ||
+            !std::is_same<typename decltype( permute )::traits::execution_space,
+                          Kokkos::Cuda>::value &&
+                std::is_same<typename decltype( permute )::value_type,
+                             size_t>::value,
+#else
+        std::is_same<typename decltype( permute )::value_type, size_t>::value,
+#endif
+        "Something change in Kokkos!?" );
+
+    // Here we chose to cast the unsigned to size_t because the size of size_t
+    // is guaranteed to coincide with the pointer size.  This is essential if
+    // we are going to convert back-and-forth with reinterpret_cast.  cast() is
+    // a no-op if the the value type in the view is already size_t.
+    return cast<size_t>( permute );
 }
 
 template <typename DeviceType>
 void TreeConstruction<DeviceType>::initializeLeafNodes(
-    Kokkos::View<int const *, DeviceType> indices,
+    Kokkos::View<size_t const *, DeviceType> indices,
     Kokkos::View<Box const *, DeviceType> bounding_boxes,
     Kokkos::View<Node *, DeviceType> leaf_nodes )
 {
     auto const n = leaf_nodes.extent( 0 );
     DTK_REQUIRE( indices.extent( 0 ) == n );
     DTK_REQUIRE( bounding_boxes.extent( 0 ) == n );
+    static_assert( sizeof( typename decltype( indices )::value_type ) ==
+                       sizeof( Node * ),
+                   "Encoding leaf index in pointer to child is not safe if the "
+                   "index and pointer types do not have the same size" );
     Kokkos::parallel_for(
         DTK_MARK_REGION( "initialize_leaf_nodes" ),
         Kokkos::RangePolicy<ExecutionSpace>( 0, n ), KOKKOS_LAMBDA( int i ) {
             leaf_nodes( i ).bounding_box = bounding_boxes( indices( i ) );
-            leaf_nodes( i ).children = {nullptr, nullptr};
+            leaf_nodes( i ).children = {
+                nullptr, reinterpret_cast<Node *>( indices( i ) )};
```

Comment:
So the main change here is now that the child pointer is also the leaf index?

Reply:
But the Morton indices are still composed to build the tree and create a permutation vector for sorting, correct?

Reply:
The pointer to children of a leaf node were null, now we use one of them to store the permutation index.

Nothing has changed as far as the Morton order is concerned.